pubsub/pubsub_local.go

```diff
@@ -44,59 +46,52 @@ func (ps *LocalPubSub) Stop() error {
 func (ps *LocalPubSub) Close() {
 	ps.mut.Lock()
 	defer ps.mut.Unlock()
-	for _, sub := range ps.subs {
-		sub.Close()
+	close(ps.done)
+	for _, subs := range ps.topics {
+		for i := range subs {
+			subs[i].cb = nil
+		}
 	}
-	ps.subs = nil
+	ps.topics = make(map[string][]*LocalSubscription, 0)
 }
 
-func (ps *LocalPubSub) ensureTopic(topic string) chan string {
+func (ps *LocalPubSub) ensureTopic(topic string) {
 	if _, ok := ps.topics[topic]; !ok {
-		ps.topics[topic] = make(chan string, 100)
+		ps.topics[topic] = make([]*LocalSubscription, 0)
 	}
-	return ps.topics[topic]
 }
 
 func (ps *LocalPubSub) Publish(ctx context.Context, topic, message string) error {
-	ps.mut.RLock()
-	ch := ps.ensureTopic(topic)
-	ps.mut.RUnlock()
-	select {
-	case ch <- message:
-	case <-ctx.Done():
-		return ctx.Err()
+	ps.mut.Lock()
+	defer ps.mut.Unlock()
+	ps.ensureTopic(topic)
+	for _, sub := range ps.topics[topic] {
+		// don't wait around for slow consumers
+		if sub.cb != nil {
+			go sub.cb(message)
```

Comment:
I think currently if a caller calls `Publish` after `Close`, `sub.cb(message)` will panic. We probably need to check whether `ps.done` is closed before calling the callbacks?

Reply:
That's why I set sub.cb to nil in the Close() functions.